### ISSUE TITLE
New lyric sources pack for ESLyric

### ISSUE THREAD

**Issue by TT-ReBORN** [ESLyric/feedback#115]
Hi ohyeah,

I have wrote some new ESLyric sources and refactored and fixed your original lyric sources scripts.
I will add them to my Georgia-ReBORN theme and hope you will also add them officially to your ESLyric source repo.

**Here is the new ESLyric source pack:**
[eslyric_source_pack.zip](https://github.com/ESLyric/feedback/files/12642841/eslyric_source_pack.zip)

### Installation:
Go to foobar's `Preferences` > `Tools` > `ESLyric` < `Lyric Option` > `Lyric Sources`
At the top right corner click on `Get More...` > `Get online script` > Download everything
Go to your `foobar2000\profile\eslyric-data\scripts\searcher` and replace all files from the [eslyric_source_pack.zip](https://github.com/ESLyric/feedback/files/12642841/eslyric_source_pack.zip)

---

### Added new sources:
- absolutelyrics.js
- bandcamp.js
- chartlyrics.js
- darklyrics.js
- elyrics.js
- letras.js
- lyricalnonsense.js (will only fetch Japanese lyrics)
- lyricalnonsense_en.js (will only fetch English lyrics)
- lyricsmania.js
- lyricsmode.js
- lyricstranslate.js
- netease_en.js (will only fetch English lyrics)
- oldielyrics.js
- plyrics.js
- songlyrics.js
- stlyrics.js

### Refactored:
All original script files have been refactored and cleaned:
- azlyrics.js
- genius.js
- kugou.js
- minilyrics.js
- musixmatch.js
- netease.js
- qqmusic.js

### Changed:
Added better name descriptions for the users that shows if the source has synced or unsynced lyrics

### Fixed:
AZLyrics - fixed search query and its results when artist names start with "A" and "The". 
In the old script the search results were not found.

---

I have tested all lyric sources and they should work properly but there is a note for you in the bandcamp.js:
```
	// ! Tell ohyeah to implement meta.comment ( %comment% ), meta.label ( %label% ) and meta.publisher ( %publisher% ),
	// ! then we can use this to improve bandcamp search results:
	// const commentUrl = getUrlFromTags(meta.comment);
	// const labelUrl = getUrlFromTags(meta.label);
	// const publisherUrl = getUrlFromTags(meta.publisher);

	// const url =
	// 	commentUrl !== '' ? `https://${commentUrl.replace(/-/g, '')}.bandcamp.com/track/${title}` :
	// 	labelUrl !== '' ? `https://${labelUrl.replace(/-/g, '')}.bandcamp.com/track/${title}` :
	// 	publisherUrl !== '' ? `https://${publisherUrl.replace(/-/g, '')}.bandcamp.com/track/${title}` :
	// 	`https://${artist.replace(/-/g, '')}.bandcamp.com/track/${title}`;
```

**To explain:** 
bandcamp search query only works if the artist name is the same as url subdomain.
But there are artist that do not have their exact bandname in the bandcamp subdomain.
Many bandcamp music files have their URL to their bandcamp site in either the `%comment%`, `%label%` or `%publisher%`.
So if you could implement `meta.comment`, `meta.label`, `meta.publisher` we could use the embedded url to have better search results.
It will search for the url in these tags via `getUrlFromTag(tag)` func and use this url instead. I think this would work, but it needs to be tested...

---

You can also close this issue here https://github.com/ESLyric/feedback/issues/76, thanks!
Hope your doing fine.

-Tom

**Comment by ESLyric** [comment]
Thanks!! I'll take a look when I have time.
Maybe I'll provide another interface later which can be called in main thread, so we can do any title format things.

**Comment by TT-ReBORN** [comment]
@ESLyric,

Hi ohyeah, 
here are the refactored lyric parser: [parser_refactored.zip](https://github.com/ESLyric/feedback/files/12700585/parser_refactored.zip)

That is all for now, cheers.

-Tom

**Comment by ESLyric** [comment]
updated. now you can download from the online fetcher.

**Comment by TT-ReBORN** [comment]
Thanks ohyeah, don't forget the lyric parser js files from the [post above](https://github.com/ESLyric/feedback/issues/115#issuecomment-1731318063).

**Comment by ESLyric** [comment]
Hi, I think the original parsers are barely ok.

**Comment by TT-ReBORN** [comment]
@ESLyric,

Hi ohyeah,

here is a new ESLyric source pack 2:
[eslyric_source_pack2.zip](https://github.com/ESLyric/feedback/files/13061025/eslyric_source_pack2.zip)

Contains 4 new lyric sources:
- Lololyrics
- LyricsOnDemand
- Metallum
- Songmeanings

Did test those a bit and all seems OK. 
This is it for now ( at least from my side ), we now have more than enough ;-)

-Tom

**Comment by ESLyric** [comment]
> @ESLyric,
> 
> Hi ohyeah,
> 
> here is a new ESLyric source pack 2: [eslyric_source_pack2.zip](https://github.com/ESLyric/feedback/files/13061025/eslyric_source_pack2.zip)
> 
> Contains 4 new lyric sources:
> 
> * Lololyrics
> * LyricsOnDemand
> * Metallum
> * SongMeanings
> 
> Did test those a bit and all seems OK. This is it for now ( at least from my side ), we now have more than enough ;-)
> 
> -Tom

@TT-ReBORN ok, added to online downloader.

**Comment by TT-ReBORN** [comment]
I can find all lyrics for all the songs you have listed via the metallum searcher script. 
It just need to have the same album title as on the site, e.g:
https://www.metal-archives.com/search/advanced/searching/songs?songTitle=Suspiria&bandName=Darkwell&releaseTitle=&lyrics=&genre=#songs

If the artist, album and title are the same as on the website, it will find them.

**Comment by TT-ReBORN** [comment]
@ESLyric,

here is the last one ( deezer ) as a bonus. Tested it, but you need to add your own API key and cookie to make it work.
Now it contains only placeholders, you can decide what to do with it.

[eslyric_source_deezer.zip](https://github.com/ESLyric/feedback/files/13062122/eslyric_source_deezer.zip)

deezer has synced and unsynced lyrics, so I have add the description "Mixed".
That is all.

-Tom

**Comment by TT-ReBORN** [comment]
Yes, I have included the album meta data to improve search results accuracy. Did not account for radio streams though as I never use them with lyrics. ohyeah could implement a new callback that checks if current playback is streaming. Then we could add conditional statement to use or not use album meta in the search.

---

@ESLyric, 

just found out about `meta.rawTitle` in your Chinese docs. I have made a workaround in the lololyrics script that is sub-optimal. With `meta.rawTitle` it has much better search results. The issue is that it works on automatic search BUT when you do ESLyric `Lyric Search`, the `meta.rawTitle` will be processed. This seems like a bug, `meta.rawTitle` should always be unprocessed ( no stripping of special chars, i.e apostrophes, commas, dots etc ). Using the latest ESLyric version 0.5.4.1019

Example:
console.log(meta.rawString); outputs `God's Child`
when using Lyric Search:
console.log(meta.rawString); outputs `Gods Child`
It has stripped the apostrophe, this was just an example.

Can you please fix this first? After that, I will then send you lololyrics.js v0.2.

-Tom

**Comment by ESLyric** [comment]
I haven't seen any preprocess when manual search
![image](https://github.com/ESLyric/feedback/assets/102273569/6f7d9d52-aec1-4641-a4d0-c6bd039aec8d)

**Comment by TT-ReBORN** [comment]
Weird, I did check with original foobar and your original ESLyric and I can reproduce it.
Here is the new improved lololyrics script v0.2:
[eslyric_source_lololyrics_v0.2.zip](https://github.com/ESLyric/feedback/files/13063385/eslyric_source_lololyrics_v0.2.zip)

**Comment by TT-ReBORN** [comment]
Ok I have found the issue, when I deactivate everything here, it works:

![issue](https://github.com/ESLyric/feedback/assets/79368795/4ce0d6e1-bfec-4e2a-99ab-f37aa3db5937)

But this should be ignored for `meta.rawArtist`, `meta.rawAlbum`, `meta.rawTitle`
Don't you think?

**Comment by ESLyric** [comment]
I checked also, the `raw` version are all expected.

**Comment by TT-ReBORN** [comment]
@ESLyric, [I have posted a video in the previous post that you can see](https://github.com/ESLyric/feedback/issues/115#issuecomment-1774104690). 
It is the ESLyric option "Remove Characters" that also changes the meta.rawTitle. 
I have console.logged this also and it changes to the correct one when I deactivate "Remove Characters".

**Comment by ESLyric** [comment]
Yes, the 'Edit' field's initial values still are preprocessed by design.

**Comment by TT-ReBORN** [comment]
OK but can you disable/ignore this for `meta.rawArtist`, `meta.rawAlbum`, `meta.rawTitle` for raw meta?
These should always be raw ( makes sense ) and not preprocessed? Otherwise if meta.rawTitle is not raw but preprocessed, forget the lololyrics v.0.2 zip I have posted as it will not work and use v0.1. The choice and effort is yours...

-Tom

**Comment by ESLyric** [comment]
The preprocess fileds are used as the initial values only when manual searchin, you can edit them. For example, if you changed the Title edit field to 'xxxx', the `meta.title` and `meta.rawTitle` will be 'xxxx'. In most cases, the preprocess values should be more convenient.

**Comment by TT-ReBORN** [comment]
Yes I have understood this from the beginning, but **why** do you use the field preprocessing for `meta.rawArtist`, `meta.rawAlbum` and `meta.rawTitle`?
This does not make sense for raw meta, it should be **not preprocessed** and `meta.rawArtist`, `meta.rawAlbum` and `meta.rawTitle` should always be **RAW ( original values )** and **not processed** via the ESLyric "Field processing" settings.

So I ask again, can you only **exclude/disable** ESLyric Field processing for `meta.rawArtist`, `meta.rawAlbum` and `meta.rawTitle` but **only enable** it for `meta.artist`, `meta.album`, `meta.title`? 
I do not want that `meta.rawArtist`, `meta.rawAlbum` and `meta.rawTitle` being processed when ESLyric Field processing option `"Remove Characters"` is active and when doing manual Lyric Search as you see in the video. I do not see any problems in changing that behavior.

Thanks for understanding.

-Tom

**Comment by ESLyric** [comment]
When manully search, the `raw` metas were not related with track's metadb, they repected the input fileds only(except `album` which cannot editable), it shouldn't use the track's original metadb in case of empty filelds or wrong infos.

So actually the initial value cause the 'issue'. As I said, Use the preprocess value as the initial values would be more conventient in most cases when search manully. If you cannot fetch the result, you should try different keywords any ways.

**Comment by Lovecraft-github** [comment]
Hi there people. Great to see more lyrics sources.
I have some problems with minilyrics.js
The console spawn
`[eslyric]script search service: eval script failed, RangeError: invalid array length    at set (native)
    at concatTypedArray (Z:/home/user/.foobar2000_v2/profile/eslyric-data/scripts/searcher/minilyrics.js:147)
    at encryptBuffer (Z:/home/user/.foobar2000_v2/profile/eslyric-data/scripts/searcher/minilyrics.js:88)
    at getLyrics (Z:/home/user/.foobar2000_v2/profile/eslyric-data/scripts/searcher/minilyrics.js:33)`

Need to be disabled.

Also some
`[eslyric]TypeError: out-of-bound numeric index`

**Comment by TT-ReBORN** [comment]
@Lovecraft-github,

happens when MiniLyrics server is down. 
I have just uploaded a new version to suppress the error log nobody wants to see...
https://github.com/ESLyric/feedback/issues/115#issuecomment-1773821993

**Comment by TT-ReBORN** [comment]
@ESLyric,

OK I get it now, this was so confusing ohyeah. Sorry :( , but you could have told me from the beginning to just check console.log when changing the title in the "Lyric Search" input field, took me long to figure it out...

**Anyways thanks for the help. I have uploaded the new improved lololyrics.js here** https://github.com/ESLyric/feedback/issues/115#issuecomment-1773821993
It is in the updated eslyric_source_pack2.zip , so please update yours and also the new minilyrics 0.4.

Thanks

-Tom

**Comment by Lovecraft-github** [comment]
Error log gone.
Thanks mate

**Comment by Lovecraft-github** [comment]
One more question.
Sometimes i have "Sorry, I have no, would you please to me? Thank You." as lyric. Maybe from musixmatch (synced)? (not sure)
Is there any way to filter this?

**Comment by ESLyric** [comment]
@TT-ReBORN  Please issue a PR instead, thanks.

**Comment by TT-ReBORN** [comment]
An idea to filter out garbage lyrics would be to make an ESLyric blacklist filter feature where user can input these strings
and ESLyric could completely ignore these lyrics.

**Comment by ESLyric** [comment]
You can remove lyric lines that meet the conditions. when no lines left, the lyric would be filtered. 
For example:
![0](https://github.com/ESLyric/feedback/assets/102273569/2c6ec0c1-9682-4bf1-9e0b-bb99abde098b)
Which will remove lines begin with `Sorry`.

**Comment by Lovecraft-github** [comment]
> You can remove lyric lines that meet the conditions. when no lines left, the lyric would be filtered. For example: ![0](https://user-images.githubusercontent.com/102273569/277523333-2c6ec0c1-9682-4bf1-9e0b-bb99abde098b.jpg) Which will remove lines begin with `Sorry`.

So no line no lyric xD
Thanks

**Comment by sveakul** [comment]
TT:  The Genius v2 searcher script now returns a first line on all lyrics that shows # of contributors and song title.  ESL 1.0.2.0. (was also present in 1.0.1.8).  Examples:

7 ContributorsCrazy Mama Lyrics1,2,3,4

29 ContributorsTake Me to the River Lyrics[Verse 1]

2 ContributorsSmall Flashing Light Lyrics[Verse 1]

5 ContributorsLucy Takes a Picture Lyrics[Verse 1

5 ContributorsLucy Takes a Picture Lyrics[Verse 1]

**Comment by sveakul** [comment]
A new v3 of the genius.js searcher was supplied this week by TT, and fixes the problem described above.  Attached below.

[genius.zip](https://github.com/user-attachments/files/20418318/genius.zip)
